test/boost/aws_error_injection_test.cc

```diff
@@ -0,0 +1,147 @@
+}
+
+void test_client_upload_file(std::string_view test_name, failure_policy policy, size_t total_size, size_t memory_size) {
+    tmpdir tmp;
+    const auto file_path = tmp.path() / fmt::format("test-{}", ::getpid());
+    {
+        file f = open_file_dma(file_path.native(), open_flags::create | open_flags::wo).get();
+        auto output = make_file_output_stream(std::move(f)).get();
+        auto close_file = deferred_close(output);
+        std::string_view data = "1234567890ABCDEF";
+        // so we can test !with_remainder case properly with multiple writes
+        SCYLLA_ASSERT(total_size % data.size() == 0);
+
+        for (size_t bytes_written = 0; bytes_written < total_size; bytes_written += data.size()) {
+            output.write(data.data(), data.size()).get();
+        }
+    }
+
+    const auto object_name = fmt::format("/{}/{}-{}", "test", test_name, ::getpid());
+    register_policy(object_name, policy);
+
+    semaphore mem{memory_size};
+    auto client = s3::client::make(get_address(), make_minio_config(), mem);
+    auto client_shutdown = deferred_close(*client);
+    client->upload_file(file_path, object_name).get();
+}
+
+SEASTAR_THREAD_TEST_CASE(test_multipart_upload_file_success) {
+    const size_t part_size = 5_MiB;
+    const size_t remainder_size = part_size / 2;
+    const size_t total_size = 4 * part_size + remainder_size;
+    const size_t memory_size = part_size;
+    BOOST_REQUIRE_NO_THROW(test_client_upload_file(seastar_test::get_name(), failure_policy::SUCCESS, total_size, memory_size));
+}
+
+SEASTAR_THREAD_TEST_CASE(test_multipart_upload_file_retryable_success) {
+    const size_t part_size = 5_MiB;
+    const size_t remainder_size = part_size / 2;
+    const size_t total_size = 4 * part_size + remainder_size;
+    const size_t memory_size = part_size;
+    BOOST_REQUIRE_NO_THROW(test_client_upload_file(seastar_test::get_name(), failure_policy::RETRYABLE_FAILURE, total_size, memory_size));
+}
+
+SEASTAR_THREAD_TEST_CASE(test_multipart_upload_file_failure) {
+    const size_t part_size = 5_MiB;
+    const size_t remainder_size = part_size / 2;
+    const size_t total_size = 4 * part_size + remainder_size;
+    const size_t memory_size = part_size;
+    BOOST_REQUIRE_EXCEPTION(test_client_upload_file(seastar_test::get_name(), failure_policy::NONRETRYABLE_FAILURE, total_size, memory_size),
+                            storage_io_error,
+                            [](const storage_io_error& e) { return e.code().value() == EIO; });
+}
+
+void do_test_client_multipart_upload(failure_policy policy) {
+    const sstring name(fmt::format("/{}/testobject-{}", "test", ::getpid()));
+    register_policy(name, policy);
+    testlog.info("Make client");
+    semaphore mem(16 << 20);
+    auto cln = s3::client::make(get_address(), make_minio_config(), mem);
+    auto close_client = deferred_close(*cln);
+
+    testlog.info("Upload object");
+    auto out = output_stream<char>(cln->make_upload_sink(name));
```

Comment:
For the followup -- also include jumbo sink too. It uploads  one or more intermediate files (called pieces) and somehow needs to test failure handling of both -- piece and target object

Reply:
Will do